extensions/neo4j/runtime/src/main/java/io/quarkus/neo4j/runtime/JTAAwareDriver.java

```diff
@@ -0,0 +1,160 @@
+
+import java.util.concurrent.CompletionStage;
+
+import javax.transaction.Status;
+import javax.transaction.Synchronization;
+import javax.transaction.TransactionSynchronizationRegistry;
+
+import org.neo4j.driver.AccessMode;
+import org.neo4j.driver.Driver;
+import org.neo4j.driver.Metrics;
+import org.neo4j.driver.Session;
+import org.neo4j.driver.SessionConfig;
+import org.neo4j.driver.async.AsyncSession;
+import org.neo4j.driver.reactive.RxSession;
+import org.neo4j.driver.types.TypeSystem;
+import org.neo4j.driver.util.Experimental;
+
+import io.quarkus.arc.Arc;
+
+final class JTAAwareDriver implements Driver {
+
+    private static final Object SESSION_KEY = new Object();
+    private final Driver delegate;
+
+    JTAAwareDriver(Driver delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public boolean isEncrypted() {
+        return delegate.isEncrypted();
+    }
+
+    @Override
+    public Session session() {
+
+        return this.session(SessionConfig.defaultConfig());
+    }
+
+    @Override
+    public Session session(SessionConfig sessionConfig) {
+
+        var synchronizationRegistryHandle = Arc.container().instance(TransactionSynchronizationRegistry.class);
+        if (synchronizationRegistryHandle.isAvailable()) {
+
+            var synchronizationRegistry = synchronizationRegistryHandle.get();
+            if (synchronizationRegistry.getTransactionStatus() == Status.STATUS_ACTIVE) {
+
+                var session = (Session) synchronizationRegistry.getResource(SESSION_KEY);
+                if (session != null) {
+                    return session;
+                }
+
+                return createAndRegisterJTAAwareSession(synchronizationRegistry, sessionConfig);
+            }
+        }
+
+        return this.delegate.session(sessionConfig);
+    }
+
+    Session createAndRegisterJTAAwareSession(TransactionSynchronizationRegistry synchronizationRegistry,
+            SessionConfig sessionConfig) {
+
+        var configBuilder = SessionConfig.builder();
+        // We will use write mode by default here… We have no way to determine this from the transactional annotation
+        // If this is unwanted, make sure you use @Transactional(NEVER) and create a config on you own.
+        configBuilder = configBuilder.withDefaultAccessMode(AccessMode.WRITE)
+                .withBookmarks(sessionConfig.bookmarks());
+        configBuilder = sessionConfig.database().map(configBuilder::withDatabase).orElse(configBuilder);
+        configBuilder = sessionConfig.fetchSize().map(configBuilder::withFetchSize).orElse(configBuilder);
+
+        var session = new JTAAwareSession(this.delegate.session(configBuilder.build()));
+
+        synchronizationRegistry.putResource(SESSION_KEY, session);
+        synchronizationRegistry.registerInterposedSynchronization(new Synchronization() {
+            @Override
+            public void beforeCompletion() {
+                session.commitAndClose();
```

Comment:
Sanne is right, you must not commit or rollback before the prepare phase starts, it will lead to inconsistent outcomes (eg the decision during the prepare phase might be to rollback).

Reply:
@loicmathieu and @michael-simons can you both confirm that your extensions are not intended to support more than a single resource in a transaction (since otherwise they would need to be XA aware).

Reply:
Yes, thats the intent. (Neo4j)

Reply:
Great, thanks

Reply:
Yes, MongoDB didn't support XA and we onlist only one Session object that we reuse during the transaction